src/contraction/src/pgr_contract.hpp

```diff
@@ -0,0 +1,379 @@
+	fn_contract.getRemovedV_string(removedVertices);
+	fn_contract.getPsuedoE_string(psuedoEdges);
+}
+
+
+template < class G >
+class Pgr_contract {
+public:
+	typedef typename G::V V;
+	typedef typename G::E E;
+	typedef typename G::V_i V_i;
+	typedef typename G::E_i E_i;
+	typedef typename G::EO_i EO_i;
+
+
+	void contract_to_level(
+		G &graph,
+		int64_t level);
+
+	void remove_1_degree_vertices(G &graph);
+
+	void remove_2_degree_vertices(G &graph);
+
+	void calculateDegrees(G &graph);
+
+	void getGraphName(std::ostringstream& name,int64_t level);
+
+	int64_t getGraph_string(G &graph,std::ostringstream& estring);
+	
+	void getRemovedE_string(G &graph,std::ostringstream& estring);
+
+	void getRemovedV_string(std::ostringstream& vstring);
+
+	void getPsuedoE_string(std::ostringstream& pstring);
+	
+	typedef typename std::map<V,std::deque<Edge> > removed_V;
+	typedef typename std::map<V,std::deque<Edge> >::iterator removed_V_i;
+	typedef typename std::map<int64_t , std::pair<int64_t,int64_t> > psuedo_E;
+	typedef typename std::map<int64_t , std::pair<int64_t,int64_t> >::iterator psuedo_E_i;
+	typedef typename std::map< int64_t, std::vector<V> > degree_to_V;
+	typedef typename std::map< int64_t, std::vector<V> >::iterator degree_to_V_i;
+	typedef typename std::deque<boost_edge_t>::iterator removed_E_i;
+	typedef typename std::vector<V>::iterator Q_i;
+private:
+	int64_t last_edge_id;
+	removed_V removedVertices;
+	psuedo_E psuedoEdges;
+	degree_to_V degree_to_V_map;
+
+};
+
+
+ /******************** IMPLEMENTATION ******************/
+
+template < class G >
+void
+Pgr_contract< G >::calculateDegrees(G &graph) {
+	EO_i out, out_end;
+	V_i vi;
+	for (vi = vertices(graph.graph).first; vi != vertices(graph.graph).second; ++vi) 
+	{
+		graph.graph[(*vi)].degree=boost::out_degree(*vi,graph.graph);
+		degree_to_V_map[graph.graph[(*vi)].degree].push_back(*vi);
+	}
+}
+
+template < class G >
+void
+Pgr_contract< G >::getGraphName(std::ostringstream& name,int64_t level)
+{
+	name << "contracted_graph_" << level;
+}
+
+template < class G >
+void
+Pgr_contract< G >::contract_to_level(G &graph,int64_t level)
+{
+	switch(level)
+	{
+		case 0:
+		remove_1_degree_vertices(graph);
+		break;
+		case 1:
+		//remove_2_degree_vertices(graph);
+		default:
+		// do nothing
+		;
+	}
+}
+
+template < class G >
+void
+Pgr_contract< G >::remove_1_degree_vertices(G &graph) {
+	EO_i out,out_end;
+	V front=degree_to_V_map[1].front();
+	int64_t frontid=-1;
+	frontid=graph.graph[front].id;
+	std::vector<V> one_degree_vertices=degree_to_V_map[1];
+	degree_to_V_i it;
+	//cout << "1 degree vertices " << one_degree_vertices.size()  << endl;
+	while (degree_to_V_map[1].size()>0)
+	{
+		//cout << "Front " << graph[front].id << endl;
+		for (boost::tie(out, out_end) = out_edges(front, graph.graph);
+			out != out_end; ++out) 
+		{
+			V s=source(*out, graph.graph);
+			V t=target(*out, graph.graph);
+			int source_id=graph.graph[s].id;
```

Comment:
int64_t
id's in osm data for example are using bigint
but better is 
auto source_id =  .....